The following questions concern a small snapshot of top-of-book data in futures markets
and ask you to consider implied orders and trades. Without showing all contracts that imply
into these markets and without showing the market-by-order data and knowing the matching algorithms,
it is impossible to say with certainty what limit orders get filled. Instead, these questions
ask whether it is possible to get filled. You do not need to determine actual fills.

You should show your work, but it does not need to be code. You _can_ code a solution, but you
could simply present your answer in a markdown cell or write out your solutions in another document.

In [1]:
import pandas as pd

top_of_book = pd.DataFrame(
    [
        {
            "symbol": "SR3H6",
            "bid_sz_00": 431,
            "bid_px_00": 96.6,
            "ask_px_00": 96.605,
            "ask_sz_00": 2190,
        },
        {
            "symbol": "SR3U7",
            "bid_sz_00": 5563,
            "bid_px_00": 96.96,
            "ask_px_00": 96.965,
            "ask_sz_00": 52,
        },
        {
            "symbol": "SR3H6-SR3U7",
            "bid_sz_00": 7,
            "bid_px_00": -37,
            "ask_px_00": -36,
            "ask_sz_00": 7,
        },
    ]
)
top_of_book

,symbol,bid_sz_00,bid_px_00,ask_px_00,ask_sz_00
0,SR3H6,431,96.60,96.605,2190
1,SR3U7,5563,96.96,96.965,52
2,SR3H6-SR3U7,7,-37.00,-36.000,7


1. What are the bid and ask, price and size, in the spread market implied IN by the outright legs?

Let $H = SR3H6$, $U = SR3U7$, and the spread is
$
S = H - U.
$

- To buy the spread: buy $H$ at ask, sell $U$ at bid  
  $
  S_{\text{implied bid}} = A_H - B_U = 96.605 - 96.960 = -0.355 \;(\approx -35.5).
  $

  Size:
  $
  \min(2190,\ 5563) = 2190.
  $

- To sell the spread: sell $H$ at bid, buy $U$ at ask  
  $
  S_{\text{implied ask}} = B_H - A_U = 96.60 - 96.965 = -0.365 \;(\approx -36.5).
  $

  Size:
  $
  \min(431,\ 52) = 52.
  $

- Implied bid in $SR3H6 - SR3U7$: $-35.5$ for 2190  
- Implied ask in $SR3H6 - SR3U7$: $-36.5$ for 52

---


2. If a market order sells 5 spreads, what price will trade? Could it be implied in the legs?

Available bids in the spread:
- Direct bid: $-37.0$ @ $7$  
- Implied bid: $-35.5$ @ $2190$

A seller hits the highest bid:
- Best bid is $-35.5$  
- Size at $-35.5$ is 2190, so 5 lots fit.


- Fill price: $-35.5$ for 5 spreads  
- Necessarily implied: there is no direct order at $-35.5$, so the trade must come from the legs (sell $U$ at $96.96$ bid and buy $H$ at $96.605$ ask).

The spread seller benefits from tight outright markets that create a “hidden” higher bid via implied-in liquidity.

---


3. If a market order buys 5 spreads, what price will trade? Could it be implied in the legs?

Available asks in the spread:
- Direct ask: $-36.0$ @ $7$  
- Implied ask: $-36.5$ @ $52$

A buyer hits the lowest ask:
- The buyer prefers $-36.5$.  
- Size at $-36.5$ is 52, so 5 lots fit.


- Fill price: $-36.5$ for 5 spreads  
- Necessarily implied: there is no direct ask at $-36.5$, so the trade must be implied from the legs (buy $U$ at $96.965$ ask and sell $H$ at $96.60$ bid).

A negative spread means “cheaper” asks are more negative. The buyer gets a strictly better price from implied-in liquidity than from the quoted spread book.

---

In [34]:
top_of_book = pd.DataFrame(
    [
        {
            "symbol": "ZSX5",
            "bid_sz_00": 23,
            "bid_px_00": 1008.5,
            "ask_px_00": 1008.75,
            "ask_sz_00": 46,
        },
        {
            "symbol": "ZSF6",
            "bid_sz_00": 72,
            "bid_px_00": 1026.25,
            "ask_px_00": 1026.5,
            "ask_sz_00": 84,
        },
        {
            "symbol": "ZSH6",
            "bid_sz_00": 16,
            "bid_px_00": 1041.75,
            "ask_px_00": 1042,
            "ask_sz_00": 8,
        },
        {
            "symbol": "ZSX5-ZSF6",
            "bid_sz_00": 4141,
            "bid_px_00": -18,
            "ask_px_00": -17.75,
            "ask_sz_00": 2187,
        },
        {
            "symbol": "ZSF6-ZSH6",
            "bid_sz_00": 348,
            "bid_px_00": -15.5,
            "ask_px_00": -15.25,
            "ask_sz_00": 1167,
        },
    ]
)
top_of_book

,symbol,bid_sz_00,bid_px_00,ask_px_00,ask_sz_00
0,ZSX5,23,1008.50,1008.75,46
1,ZSF6,72,1026.25,1026.50,84
2,ZSH6,16,1041.75,1042.00,8
3,ZSX5-ZSF6,4141,-18.00,-17.75,2187
4,ZSF6-ZSH6,348,-15.50,-15.25,1167


4. What is the best bid and offer, price and size,
implied OUT in the `ZSF6` market by each of the spreads and other legs?

Let $X = ZSX5$, $F = ZSF6$, $H = ZSH6$.

Spreads:
- $S_1 = X - F = ZSX5 - ZSF6$  
- $S_2 = F - H = ZSF6 - ZSH6$

From $S_1 = X - F$ we have $F = X - S_1$.

- Implied bid in $F$ from $X$ and $S_1$ (synthetic seller of $F$ = short $F$):  
  short $F$ = short $X$ + long $S_1$: 
  $
  F_{\text{bid}, S_1} = B_X - A_{S_1} = 1008.50 - (-17.75) = 1026.25.
  $

  Size: 
  $
  \min(23,\ 2187) = 23.
  $

- Implied ask in $F$ from $X$ and $S_1$ (synthetic buyer of $F$ = long $F$):  
  long $F$ = long $X$ + short $S_1$: 
  $
  F_{\text{ask}, S_1} = A_X - B_{S_1} = 1008.75 - (-18.00) = 1026.75.
  $
  
  Size:
  $
  \min(46,\ 4141) = 46.
  $

From $S_2 = F - H$ we have $F = S_2 + H$.

- Implied bid in $F$ from $H$ and $S_2$ (short $F$):  
  short $F$ = short $S_2$ + short $H$: 
  $
  F_{\text{bid}, S_2} = B_{S_2} + B_H = -15.50 + 1041.75 = 1026.25.
  $
  
  Size:
  $
  \min(348,\ 16) = 16.
  $

- Implied ask in $F$ from $H$ and $S_2$ (long $F$):  
  long $F$ = long $S_2$ + long $H$: 
  $
  F_{\text{ask}, S_2} = A_{S_2} + A_H = -15.25 + 1042.00 = 1026.75.
  $
  
  Size:
  $
  \min(1167,\ 8) = 8.
  $


- From $ZSX5 - ZSF6$:  
  - Implied bid in $ZSF6$: $1026.25$ for 23  
  - Implied ask in $ZSF6$: $1026.75$ for 46

- From $ZSF6 - ZSH6$:  
  - Implied bid in $ZSF6$: $1026.25$ for 16  
  - Implied ask in $ZSF6$: $1026.75$ for 8

Aggregating implied-out liquidity (ignoring queueing):

- Total implied bid at $1026.25$: $23 + 16 = 39$  
- Total implied ask at $1026.75$: $46 + 8 = 54$

Direct $ZSF6$ is $1026.25$ bid (72) / $1026.50$ ask (84), so:

- Bid: implied and direct are at the same price.  
- Ask: direct ($1026.50$) is better (lower) than any implied ask ($1026.75$).

---

5. Suppose a market order sells 5 `ZSF6`. At what price is the ordered filled? Is the fill potentially
an implied tade?

Best bids in $ZSF6$:
- Direct: $1026.25$ @ $72$  
- Implied from $ZSX5 - ZSF6$: $1026.25$ @ $23$  
- Implied from $ZSF6 - ZSH6$: $1026.25$ @ $16$

So the best bid price is $1026.25$:
- Fill price: $1026.25$ for 5 contracts  
- Fill may be implied or direct (or a mix): at $1026.25$ there is both outright and implied-out depth, so from the snapshot alone we cannot tell which specific counterparty type fills the trade.

The spreads are supporting the same best bid as the outright, deepening the true executable liquidity.

---


6. Suppose a market order buys 5 `ZSF6`. At what price is the ordered filled? Is the fill potentially
an implied tade?

Best asks in $ZSF6$:
- Direct ask: $1026.50$ @ $84$  
- Implied-out asks: $1026.75$ (worse) from both spreads

The lowest ask is the direct $1026.50$.

- Fill price: $1026.50$ for 5 contracts  
- Not implied: the trade executes fully against the direct outright ask, because all implied asks are at a worse price ($1026.75$).

In this snapshot, spread markets help on the bid but not on the ask for $ZSF6$, so buyers only see improvement from the outright book.
